### PR TITLE
[REFACTOR] Streamline Function Attr interface.

### PR DESCRIPTION
**Pull request by tqchen**
There has been quite a few recent changes that depends heavily on
the function attr interface. This PR streamlines that interface by introducing
two APIs that covers most of the usages.

Thanks to @jroesch for discussing the API naming convention.

- GetAttr which gets a typed object for a given key
  - HasNonzeroAttr is a quick helper that calls GetAttr to quickly check an attribute
- WithAttr that creates a new function object with the given attr
  - The API comes with copy on write optimization to avoid multiple copies
  - We deliberately pick the prefix With(instead of Set) to indicate this
    function does not mutate the original input.

On the python side:
- We allow read access via func.attrs (which is a DictAttr)
    e.g. `func.attrs["__param__"]` 
- func.with_attrs to create a new instance with updated attrs.

We also get rid of the small wrapper functions and make sure the API centered around
the GetAttr and HasNonzeroAttr interface.

This PR also changes the function construction to follow the new convention.

cc @jroesch @zhiics @mbaret

### CODE REVIEW

**Comment by tqchen** [comment]
NOTE: I have not removed the UseDefaultCompiler member function, however, we should remove that in a followup PR by simply setting the default compiler to None and we could use `GetAttr<StringImm>(attr::kCompiler).defined()` to check whether the customization is defined.

StringImm should be changed to String once #4628  is merged

cc @comaniac @zhiics

File: include/tvm/ir/function.h [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+   * This function can be used to check whether an optional
+   * attribute mark(e.g. inline) exists.
+   *
+   * \param attr_key The key to the attribute.
+   * \return The check result.
+   *
+   * \code
+   *
+   *  void HasNonzeroAttrExample(const BaseFunc& f) {
+   *    if (f->HasNonzeroAttr(attr::Inline)) {
```
Comment (zhiics) [inline include/tvm/ir/function.h]: kInline

File: include/tvm/relay/function.h [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+ *        the attribute value key with the value.
+ *
+ * \param func The input function.
+ * \param attr_key The attribute key.
+ * \param attr_value The value attribute value.
+ *
+ * \returns The new function with updated attributes.
+ *
+ * \note This function performs copy on write optimization for func.
+ *       If we move an uniquely referenced func into WithAttr,
```
Comment (zhiics) [inline include/tvm/relay/function.h]: s/an/a

**Comment by tqchen** [comment]
thanks @zhiics please take another look

File: include/tvm/ir/function.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/ir/expr.h
+ * \brief Base expr nodes in TVM.
```
Comment (siju-samuel) [inline include/tvm/ir/function.h]: Update file & brief here

File: src/ir/function.cc [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/tvm/ir/function.cc
```
Comment (siju-samuel) [inline src/ir/function.cc]: src/ir/function.cc

File: src/relay/ir/function.cc [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/tvm/relay/ir/function.cc
```
Comment (siju-samuel) [inline src/relay/ir/function.cc]: same as above

**Comment by zhiics** [comment]
Thanks @tqchen, @ZihengJiang, @siju-samuel

**Pull request merged**